library/alloc/src/collections/fixed_queue/mod.rs

```diff
@@ -0,0 +1,288 @@
+    /// is popped and returned.
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub fn push(&mut self, item: T) -> Option<T> {
+        // 'pop' first
+        let overwritten = self.buffer[self.tail].take();
+        // overwrite head/tail element
+        self.buffer[self.tail] = Some(item);
+        // shift tail with 'push'
+        self.tail = (self.tail + 1) % N;
+        if overwritten.is_some() {
+            // shift head ptr on collision
+            self.head = (self.head + 1) % N;
+        } else {
+            // increase len if no collision
+            self.len += 1;
+        }
+        return overwritten;
+    }
+
+    /// Removes and returns the oldest element from the queue.
+    #[inline]
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub fn pop(&mut self) -> Option<T>
+    where
+        Option<T>: Copy,
+    {
+        if self.len == 0 {
+            return None;
+        }
+        let popped = self.buffer[self.head].take();
+        self.head = (self.head + 1) % N;
+        self.len -= 1;
+        popped
+    }
+
+    /// Converts the queue into its array equivalent.
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub fn to_option_array(self) -> [Option<T>; N]
+    where
+        Option<T>: Copy,
+    {
+        let mut arr: [Option<T>; N] = [None; N];
+        for i in 0..N {
+            arr[i] = self.buffer[(self.head + i) % N];
+        }
+        arr
+    }
+
+    /// Converts the queue into its vec equivalent.
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub fn to_vec(self) -> Vec<T>
+    where
+        T: Copy,
+    {
+        let mut vec: Vec<T> = Vec::new();
+        for i in 0..N {
+            if let Some(e) = self.buffer[(self.head + i) % N] {
+                vec.push(e);
+            }
+        }
+        vec
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T, const N: usize> From<[T; N]> for FixedQueue<T, N> {
+    /// Creates a FixedQueue from a fixed size array.
+    fn from(array: [T; N]) -> Self {
+        FixedQueue::with(array.map(Some), 0, 0, N)
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T: Copy, const N: usize> From<&[T; N]> for FixedQueue<T, N> {
+    /// Creates a FixedQueue from a fixed size slice.
+    fn from(array: &[T; N]) -> Self {
+        FixedQueue::with(array.map(Some), 0, 0, N)
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T: Copy, const N: usize> From<&[T]> for FixedQueue<T, N> {
+    /// Creates a FixedQueue from an unsized slice. Copies a maximum of N
+    /// elements of the slice, and a minimum of the slice length into the
+    /// queue. {[0, 0], 0} - [0, 0]
+    fn from(array: &[T]) -> Self {
+        let mut buf: [Option<T>; N] = [None; N];
+        let length = N.min(array.len());
+        for i in 0..length {
+            buf[i] = Some(array[i]);
+        }
+        FixedQueue::with(buf, 0, array.len() - 1, length)
+    }
+}
```

Comment:
And what if N is smaller than the number of elements? This is the problem with code that uses *indexing* constantly: it does not take advantage of Rust's ability to detect incorrect usage of types. `for` loops should be over elements in the array or slice, ideally never over indices.